stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+    /// unconfirmed tenure.  Only do this if (1) the sortition DB is at the burnchain tip and (2)
+    /// all of our wanted tenures are marked as either downloaded or complete.
+    ///
+    /// To fully determine if it's appropriate to download unconfirmed tenures, the caller should
+    /// additionally ensure that there are no in-flight confirmed tenure downloads.
+    ///
+    /// This method is static to facilitate testing.
+    pub(crate) fn need_unconfirmed_tenures<'a>(
+        nakamoto_start_block: u64,
+        burnchain_height: u64,
+        sort_tip: &BlockSnapshot,
+        completed_tenures: &HashSet<ConsensusHash>,
+        wanted_tenures: &[WantedTenure],
+        prev_wanted_tenures: &[WantedTenure],
+        tenure_block_ids: &HashMap<NeighborAddress, AvailableTenures>,
+        pox_constants: &PoxConstants,
+        first_burn_height: u64,
+        inventory_iter: impl Iterator<Item = &'a NakamotoTenureInv>,
+        blocks_db: NakamotoStagingBlocksConnRef,
+    ) -> bool {
+        if sort_tip.block_height < burnchain_height {
+            test_debug!(
+                "sort_tip {} < burn tip {}",
+                sort_tip.block_height,
+                burnchain_height
+            );
+            return false;
+        }
+
+        if wanted_tenures.len() == 0 {
+            test_debug!("No wanted tenures");
+            return false;
+        }
+
+        if prev_wanted_tenures.len() == 0 {
+            test_debug!("No prev wanted tenures");
+            return false;
+        }
+
+        // there are still confirmed tenures we have to go and get
+        if Self::have_unprocessed_tenures(
+            pox_constants
+                .block_height_to_reward_cycle(first_burn_height, nakamoto_start_block)
+                .expect("FATAL: nakamoto starts before system start"),
+            completed_tenures,
+            prev_wanted_tenures,
+            tenure_block_ids,
+            pox_constants,
+            first_burn_height,
+            inventory_iter,
+        ) {
+            test_debug!("Still have unprocessed tenures, so we don't need unconfirmed tenures");
+            return false;
+        }
+
+        // see if we need any tenures still
+        let mut need_tenure = false;
+        for (_naddr, available) in tenure_block_ids.iter() {
+            for wt in wanted_tenures.iter() {
+                if !available.contains_key(&wt.tenure_id_consensus_hash) {
+                    continue;
+                }
+                if completed_tenures.contains(&wt.tenure_id_consensus_hash) {
+                    continue;
+                }
+                if !wt.processed {
+                    test_debug!(
+                        "Still need tenure {} from {}",
+                        &wt.tenure_id_consensus_hash,
+                        _naddr
+                    );
+                    need_tenure = true;
+                    break;
```

Comment:
This break only breaks the inner loop, when I think you want to break the outer loop too. The ordering of this loop is also somewhat inefficient `wt.processed` could be checked `tenure_block_ids.len()` times for the same `wt` and the same goes for `completed_tenures.contains`. Nested `find`s might be more confusing, but even one `find` would make things easier:

```rust
for wt in wanted_tenures.iter() {
    if completed_tenures.contains(&wt.tenure_id_consensus_hash) {
        continue;
    }
    let is_available = tenure_block_ids
        .iter()
        .any(|(_, available)| available.contains_key(&wt.tenure_id_consensus_hash));
    if is_available && !wt.processed {
         return true;
    }
}

false
```

Reply:
Sure